kaggle_environments/agent.py

```diff
@@ -12,51 +12,78 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+import json
 import os
 import requests
-import json
-from multiprocessing import Manager, Process
+import sys
+from io import StringIO
 from time import time
-import uuid
-from .errors import DeadlineExceeded
-from .utils import get_exec, has, is_url, read_file, structify
+from urllib.parse import urlparse
+from .errors import DeadlineExceeded, InvalidArgument
+from .utils import read_file, structify
+
+
+def is_url(url):
+    try:
+        result = urlparse(url)
+        return all([result.scheme, result.netloc])
+    except ValueError:
+        return False
+
+
+def get_last_callable(raw, fallback=None):
+    orig_out = sys.stdout
+    buffer = StringIO()
+    sys.stdout = buffer
+
+    try:
+        code_object = compile(raw, "<string>", "exec")
+        env = {}
+        exec(code_object, env)
+        sys.stdout = orig_out
+        output = buffer.getvalue()
+        if output:
+            print(output)
+        return [v for v in env.values() if callable(v)][-1]
+    except Exception as e:
+        sys.stdout = orig_out
+        output = buffer.getvalue()
+        if output:
+            print(output)
+        if fallback is not None:
+            return fallback
+        raise InvalidArgument("Invalid raw Python: " + repr(e))
 
 
-def build_agent(raw, environment, debug=False):
+def build_agent(raw, environment):
+    if raw in environment.agents:
+        return environment.agents[raw]
+
     # Already callable.
     if callable(raw):
         return raw
 
     # Not a string, static action.
-    if not has(raw, str):
+    if not isinstance(raw, str):
         return lambda: raw
 
     # A URL and will be initialized on the calling server.
     if is_url(raw):
-        def url_agent(o, c, r, i):
```

Comment:
As far as I can tell the reward and info parameters are never actually used for remote agents so I dropped these.